modalc-lib/modal-vector.rkt

```diff
@@ -45,10 +43,54 @@
                                     ((vector-ref inner-ctc-proj/blame index) value neg-party)]
                                    [else value]))
                            (λ (vec index value)
-                             (cond [(should-apply-ctc?/set vec)
+                             (cond [(should-apply-ctc?/set (list index value))
                                     ((vector-ref inner-ctc-proj/blame index) value neg-party)]
                                    [else value])))))))
 
+(define (modal-vectorof should-apply-ctc? . inner-ctc)
+  (apply modal-vectorof* inner-ctc #:ref/set should-apply-ctc?))
+
+(define (modal-vectorof* #:ref [ref-ctc? #f]
+                        #:set [set-ctc? #f]
+                        #:ref/set [all-ctc? #f]
+                        . inner-ctc)
+  
+  ;; error when no mode exists
+  (unless (or ref-ctc?
+              set-ctc?
+              all-ctc?)
+    (raise-user-error 'modal-vector/c* "No mode supplied"))
+
+  ;; create ref and set modes
+  (define should-apply-ctc?/ref (or ref-ctc?
+                                    all-ctc?
+                                    mode:never))
+  (define should-apply-ctc?/set (or set-ctc?
+                                    all-ctc?
+                                    mode:never))
+  (define inner-ctc-proj
+    (apply contract-late-neg-projection inner-ctc))
+
+  (make-contract
+   #:name `(modal-vector/c ,(contract-name inner-ctc-proj))
+   #:late-neg-projection
+   (λ (blame)
+     (define inner-ctc-proj/blame (inner-ctc-proj blame))
+     (λ (val neg-party)
+       (impersonate-vector val
+                           (λ (vec index value)
+                             (cond [(should-apply-ctc?/ref (list index value))
+                                    (for/vector ([v vec])
+                                      (inner-ctc-proj/blame v neg-party))
+                                    value]
```

Comment:
Why is this different than the checking `modal-vector/c` does -- on just the relevant index?

Reply:
See if you can abstract the implementation of `modal-vector/c` in such a way that you can reuse it here.
One idea to do that: write a `make-vector-projection` function that takes as argument a two functions, which given an index return either the `ref` contract or the `set` contract for that index.